pkg/apk/auth/chainguard.go

```diff
@@ -0,0 +1,49 @@
+package auth
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"chainguard.dev/sdk/sts"
+	"google.golang.org/api/idtoken"
+)
+
+// NewChainguardIdentityAuth returns an Authenticator that authorizes
+// requests as the given assumeable identity.
+//
+// The identity is a UIDP of a Chainguard Identity.
+// Issuer is usually https://issuer.enforce.dev.
+// Audience is usually https://apk.cgr.dev.
+func NewChainguardIdentityAuth(identity, issuer, audience string) Authenticator {
+	return authenticator{
+		id:  identity,
+		iss: issuer,
+		aud: audience,
+	}
+}
+
+type authenticator struct {
+	id, iss, aud string
+}
+
+func (a authenticator) AddAuth(ctx context.Context, req *http.Request) error {
+	if req.Host != strings.TrimPrefix(a.aud, "https://") {
+		return nil
+	}
+
+	ts, err := idtoken.NewTokenSource(ctx, a.iss)
+	if err != nil {
+		return err
+	}
+	tok, err := ts.Token()
+	if err != nil {
+		return err
+	}
+	ctok, err := sts.Exchange(ctx, a.iss, a.aud, tok.AccessToken, sts.WithIdentity(a.id))
+	if err != nil {
+		return err
+	}
```

Comment:
So this doesn't cache at all.  Consider constructing the `TokenSource` in the constructor and wrapping with this to get caching for the token lifetime?

https://pkg.go.dev/golang.org/x/oauth2#ReuseTokenSource

Reply:
I've switched to use a `sometimes` like we do for chainctl auth, for consistency.